### PR TITLE
Added: support to set current location from PO details and shipment details page (#238pmgz)

### PR DESCRIPTION
**Pull request by disha1202**
### Related Issues

Closes #

### Screenshots of Visual Changes before/after (If There Are Any)

**IMPORTANT NOTICE** - Remember to add changelog entry

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/hotwax/receiving#contribution-guideline)

### CODE REVIEW

File: src/store/modules/user/actions.ts [hunk 1]
```diff
@@ -52,24 +52,63 @@ const actions: ActionTree<UserState, RootState> = {
...
         emitter.emit('timeZoneDifferent', { profileTimeZone: resp.data.userTimeZone, localTimeZone});
       }
+      if(resp.data.facilities.length > 0) {
+        const payload = {
+          "inputFields": {
+            "facilityId": resp.data.facilities[0].facilityId
+          },
+          "fieldList": ["locationSeqId", "areaId", "aisleId", "sectionId", "levelId", "positionId"],
+          "entityName": "FacilityLocation",
+          "distinct": "Y",
```
Comment (ymaheshwari1) [inline src/store/modules/user/actions.ts:69]: Add viewSize.

File: src/store/modules/user/actions.ts [hunk 2]
```diff
@@ -52,24 +52,63 @@ const actions: ActionTree<UserState, RootState> = {
...
+      },
+      "viewSize": 20,
+      "fieldList": ["locationSeqId", "areaId", "aisleId", "sectionId", "levelId", "positionId"],
+      "entityName": "FacilityLocation",
+      "distinct": "Y",
+      "noConditionFind": "Y"
+    }
+
+    await dispatch('getFacilityLocations', param).then((locations: any) => {
+      user.facilityLocations = locations
```
Comment (ymaheshwari1) [inline src/store/modules/user/actions.ts:101]: I think we should not append the result in user object, and we can directly commit the data in the then part only.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/views/ShipmentDetails.vue [hunk 3]
```diff
@@ -28,27 +28,38 @@
...
+            </div>
+
+            <div class="location">
+              <LocationPopover />
+            </div>
+
+            <div class="product-count">
+              <ion-item>
+                <ion-label position="floating">{{ $t("Qty") }}</ion-label>
+                <ion-input type="number" min="0" v-model="item.quantityAccepted"></ion-input>
```
Comment (ymaheshwari1) [inline src/views/ShipmentDetails.vue:51]: Use self-closing tag.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/theme/variables.css [hunk 4]
```diff
@@ -260,6 +260,47 @@ ion-thumbnail > img {
   margin-top: 25px;
 }
 
+.scanner {
+  display: grid;
+  grid-template-columns: repeat(auto-fit, minmax(343px, 1fr));
+  gap: 8px;
```
Comment (azkyakhan) [inline src/theme/variables.css:266]: Instead of pixel values please use spacer variables where we have provided spaces.

**Review by azkyakhan — CHANGES_REQUESTED** [review]

File: src/theme/variables.css (hunk 4 above)
Comment (disha1202) [reply src/theme/variables.css]: Used spacer variables.

File: src/views/ShipmentDetails.vue (hunk 3 above)
Comment (disha1202) [reply src/views/ShipmentDetails.vue]: Used self-closing tag.

**Review by azkyakhan — APPROVED** [review]

File: src/views/PurchaseOrderDetail.vue [hunk 5]
```diff
@@ -67,13 +64,30 @@
...
-            </ion-button>
+          <div class="action border-top" v-if="item.quantity > 0">
+            <div class="receive-all-qty">
+              <ion-button @click="receiveAll(item)" slot="start" fill="outline">
+                {{ $t("Receive All") }}
+              </ion-button>
+            </div>
+
+            <div class="qty-progress">
             <ion-progress-bar :value="item.quantityAccepted/item.quantity" />
```
Comment (azkyakhan) [inline src/views/PurchaseOrderDetail.vue:75]: Improve indentation

**Review by azkyakhan — CHANGES_REQUESTED** [review]

File: src/views/PurchaseOrderDetail.vue (hunk 5 above)
Comment (disha1202) [reply src/views/PurchaseOrderDetail.vue]: Improved indentation

**Review by azkyakhan — APPROVED** [review]

File: src/views/Settings.vue [hunk 6]
```diff
@@ -68,7 +68,7 @@ export default defineComponent({
     setFacility (facility: any) {
       if(this.userProfile && this.userProfile.facilities) {
         this.userProfile.facilities.map((fac: any) => {
-          if (fac.facilityId == facility['detail'].value) {
+          if (fac.facilityId == facility['detail'].value && this.currentFacility.facilityId != facility['detail'].value) {
```
Comment (ymaheshwari1) [inline src/views/Settings.vue:71]: Add this check before the userProfile check as if the values are not equal then there is no need to loop over user facilities.
Also add a comment Todo that why we have added this check.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/views/Settings.vue (hunk 6 above)
Comment (disha1202) [reply src/views/Settings.vue]: Added the check before userProfile check and comment.

File: src/components/LocationPopover.vue [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  computed: {
+    ...mapGetters({
+      facilityLocations: 'user/getFacilityLocations',
+      currentFacilityLocation: 'user/getCurrentFacilityLocation',
+    })
+  },
+  methods: {
+    setFacilityLocation(facilityLocation: any) {
+      if (this.facilityLocations) {
+        this.facilityLocations.map((location: any) => {
```
Comment (adityasharma7) [inline src/components/LocationPopover.vue:38]: I think we should find the location and then dispatch.

**Review by adityasharma7 — APPROVED** [review]

File: src/store/modules/user/actions.ts [hunk 8]
```diff
@@ -52,13 +52,17 @@ const actions: ActionTree<UserState, RootState> = {
...
    */
-  async getProfile ( { commit }) {
+  async getProfile ( { commit, dispatch }) {
     const resp = await UserService.getProfile()
     if (resp.status === 200) {
       const localTimeZone = moment.tz.guess();
       if (resp.data.userTimeZone !== localTimeZone) {
         emitter.emit('timeZoneDifferent', { profileTimeZone: resp.data.userTimeZone, localTimeZone});
       }
+      if(resp.data.facilities.length > 0) {
```
Comment (adityasharma7) [inline src/store/modules/user/actions.ts:62]: I think we should rethink if we need to fetch this at login

File: src/store/modules/user/actions.ts [hunk 9]
```diff
@@ -67,9 +71,14 @@ const actions: ActionTree<UserState, RootState> = {
   /**
    * update current facility information
    */
-  async setFacility ({ commit }, payload) {
+  async setFacility ({ commit, dispatch }, payload) {
+    await dispatch('getFacilityLocations', payload.facility.facilityId)
     commit(types.USER_CURRENT_FACILITY_UPDATED, payload.facility);
   },
+
+  async setFacilityLocation({ commit }, payload) {
```
Comment (adityasharma7) [inline src/store/modules/user/actions.ts:79]: If this is not a global, we should discuss if we want such experience if facility location once set on shipment details or PO details page, it should be stored in state and reuse as preferred configuration

Comment (adityasharma7) [reply src/store/modules/user/actions.ts:79]: @dt2patel Any inputs?

File: src/components/LocationPopover.vue [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    IonSelectOption
+  },
+  computed: {
+    ...mapGetters({
+      facilityLocations: 'user/getFacilityLocations',
+      currentFacilityLocation: 'user/getCurrentFacilityLocation',
+    })
+  },
+  methods: {
+    setFacilityLocation(facilityLocation: any) {
```
Comment (adityasharma7) [inline src/components/LocationPopover.vue:36]:
```suggestion
    setFacilityLocation(event: any) {
```

File: src/components/LocationPopover.vue [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+      if (this.facilityLocations) {
+        const location = this.facilityLocations.find((location: any) => location.locationSeqId === facilityLocation['detail'].value)
```
Comment (adityasharma7) [inline src/components/LocationPopover.vue:38]:
```suggestion
        const facilityLocation = this.facilityLocations.find((location: any) => location.locationSeqId === facilityLocation['detail'].value)
```

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/store/modules/order/actions.ts [hunk 12]
```diff
@@ -135,11 +135,10 @@ const actions: ActionTree<OrderState, RootState> = {
             })
           })
 
-          // TODO: remove the hardcoded value, currently using harcoded locationSeqId for NotNaked catalog
           const poShipment = {
             shipment: {
               shipmentId,
-              locationSeqId: 'TLTLTLLL02'
+              locationSeqId: rootGetters['user/getCurrentFacilityLocation'].locationSeqId
```
Comment (adityasharma7) [inline src/store/modules/order/actions.ts:141]: I think we could directly access state here

File: src/store/modules/user/UserState.ts [hunk 13]
```diff
@@ -2,5 +2,7 @@ export default interface UserState {
     token: string;
     current: object | null;
     currentFacility: object | null;
+    currentFacilityLocation: object | null;
```
Comment (adityasharma7) [inline src/store/modules/user/UserState.ts:5]: We may be having already set facility location for shipments and purchase order. We need to rethink if this is to be persisted in state or we can just have the existing or first value from the list for empty case

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/store/modules/user/actions.ts (hunk 9 above)
Comment (dt2patel) [reply src/store/modules/user/actions.ts:104]: we don't need to manage a default facility location at an app level. Product facilities will already have pick primary locations and we can use those.

File: src/store/modules/shipment/mutation-types.ts [hunk 14]
```diff
@@ -1,4 +1,5 @@
 export const SN_SHIPMENT = 'shipment'
 export const SHIPMENT_LIST_UPDATED = SN_SHIPMENT + '/SEARCH_UPDATED'
 export const SHIPMENT_CURRENT_UPDATED = SN_SHIPMENT + '/CURRENT_UPDATED'
-export const SHIPMENT_CURRENT_PRODUCT_ADDED = SN_SHIPMENT + '/PRODUCT_ADDED'
\ No newline at end of file
+export const SHIPMENT_CURRENT_PRODUCT_ADDED = SN_SHIPMENT + '/PRODUCT_ADDED'
```
Comment (ymaheshwari1) [inline src/store/modules/shipment/mutation-types.ts:4]:
```suggestion
export const SHIPMENT_CURRENT_PRODUCT_ADDED = SN_SHIPMENT + '/CURRENT_PRODUCT_ADDED'
```

File: src/store/modules/user/actions.ts [hunk 15]
```diff
@@ -115,6 +118,41 @@ const actions: ActionTree<UserState, RootState> = {
   // Set User Instance Url
   setUserInstanceUrl ({ commit }, payload){
     commit(types.USER_INSTANCE_URL_UPDATED, payload)
+  },
+
+  async getFacilityLocations( { commit }, facilityId ) {
+    let resp;
+    const payload = {
+      "inputFields": {
+        "facilityId": facilityId
```
Comment (ymaheshwari1) [inline src/store/modules/user/actions.ts:127]:
```suggestion
        facilityId
```

File: src/store/modules/user/actions.ts [hunk 16]
```diff
@@ -115,6 +118,41 @@ const actions: ActionTree<UserState, RootState> = {
...
+      // Assuming we will not have more than 20 facility locations, hardcoded the viewSize value 20.
+      "viewSize": 20,
+      "fieldList": ["locationSeqId", "areaId", "aisleId", "sectionId", "levelId", "positionId"],
+      "entityName": "FacilityLocation",
+      "distinct": "Y",
+      "noConditionFind": "Y"
+    }
+    try{
+      resp = await UserService.getFacilityLocations(payload);
+      if(resp.status === 200 && resp.data.count > 0 && !hasError(resp)) {
```
Comment (ymaheshwari1) [inline src/store/modules/user/actions.ts:138]:
```suggestion
      if(resp.status === 200 && !hasError(resp) && resp.data.count > 0) {
```

File: src/views/PurchaseOrderDetail.vue [hunk 17]
```diff
@@ -229,9 +248,12 @@ export default defineComponent({
     }
   },
   mounted() {
+    if(!this.facilityLocations.length && this.currentFacility.facilityId) {
+      this.store.dispatch('user/getFacilityLocations', this.currentFacility.facilityId)
+    }
     this.store.dispatch("order/getOrderDetail", { orderId: this.$route.params.slug }).then(() => {
       this.store.dispatch('order/getPOHistory', { orderId: this.order.orderId })
-    })
+    })
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrderDetail.vue:256]: Unwanted change, please revert.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/components/LocationPopover.vue [hunk 18]
```diff
@@ -0,0 +1,53 @@
+<template>
+  <ion-chip outline>
+    <ion-icon :icon="locationOutline"/>
+    <ion-select interface="popover" :placeholder="$t('facility location')" :value="item.locationSeqId ? item.locationSeqId : facilityLocations[0].locationSeqId" @ionChange="setFacilityLocation($event)">
```
Comment (ymaheshwari1) [inline src/components/LocationPopover.vue:4]: If there are no facilityLocations then `facilityLocations[0].locationSeqId` will result in undefined error.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/components/LocationPopover.vue [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    IonSelectOption
+  },
+  computed: {
+    ...mapGetters({
+      facilityLocations: 'user/getFacilityLocations',
+    })
+  },
+  props: ['item', 'type'],
+  methods: {
+    setFacilityLocation(event: any) {
```
Comment (ymaheshwari1) [inline src/components/LocationPopover.vue:36]:
```suggestion
    setFacilityLocation(event: CustomEvent) {
```

File: src/store/modules/order/actions.ts [hunk 20]
```diff
@@ -168,12 +165,20 @@ const actions: ActionTree<OrderState, RootState> = {
...
         },
         "entityName": "ShipmentReceiptAndItem",
-        "fieldsToSelect": ["datetimeReceived", "productId", "quantityAccepted", "quantityRejected", "receivedByUserLoginId", "shipmentId"]
+        "fieldList": ["datetimeReceived", "productId", "quantityAccepted", "quantityRejected", "receivedByUserLoginId", "shipmentId", 'locationSeqId']
       }
       resp = await OrderService.fetchPOHistory(params)
       if (resp.status === 200 && !hasError(resp) && resp.data?.count > 0) {
         const poHistory = resp.data.docs;
         current.poHistory.items = poHistory;
+        const facilityLocationByProduct = poHistory?.reduce((products: any, item: any) => {
```
Comment (ymaheshwari1) [inline src/store/modules/order/actions.ts:174]: I think we don't need optional chaining here, as we already check for resp.data.count and it that check passes then it means that we will have data in resp.data.docs

File: src/store/modules/shipment/actions.ts [hunk 21]
```diff
@@ -49,7 +49,8 @@ const actions: ActionTree<ShipmentState, RootState> = {
         commit(types.SHIPMENT_CURRENT_UPDATED, { current: resp.data })
         let productIds: any = new Set();
         resp.data.items.forEach((item: any) => {
-          productIds.add(item.productId)
+          productIds.add(item.productId);
```
Comment (ymaheshwari1) [inline src/store/modules/shipment/actions.ts:52]: Unwanted change

File: src/store/modules/user/actions.ts [hunk 22]
```diff
@@ -93,6 +93,7 @@ const actions: ActionTree<UserState, RootState> = {
         await dispatch('getEComStores', { facilityId: resp.data.facilities[0].facilityId })
       }
       commit(types.USER_CURRENT_FACILITY_UPDATED, resp.data.facilities.length > 0 ? resp.data.facilities[0] : {});
+      dispatch('getFacilityLocations', resp.data.facilities.length > 0 ? resp.data.facilities[0].facilityId : '')
```
Comment (ymaheshwari1) [inline src/store/modules/user/actions.ts:96]: We always need to fetch locations for 0th facility, for now add a TODO to remove this check once adding support to not allow user login without facilities.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

**Pull request merged**